win32ss/drivers/miniport/xboxvmp/xboxvmp.c

```diff
@@ -389,25 +389,13 @@ XboxVmpMapVideoMemory(
 {
     PHYSICAL_ADDRESS FrameBuffer;
     ULONG inIoSpace = VIDEO_MEMORY_SPACE_MEMORY;
-    SYSTEM_BASIC_INFORMATION BasicInfo;
-    ULONG Length;
 
-    /* FIXME: this should probably be done differently, without native API */
     StatusBlock->Information = sizeof(VIDEO_MEMORY_INFORMATION);
 
-    FrameBuffer.u.HighPart = 0;
-    if (ZwQuerySystemInformation(SystemBasicInformation,
-                                 (PVOID)&BasicInfo,
-                                 sizeof(SYSTEM_BASIC_INFORMATION),
-                                 &Length) == NO_ERROR)
-    {
-        FrameBuffer.u.LowPart = BasicInfo.HighestPhysicalPageNumber * PAGE_SIZE;
-    }
-    else
-    {
-        ERR_(IHVVIDEO, "ZwQueryBasicInformation failed, assuming 64MB total memory\n");
-        FrameBuffer.u.LowPart = 60 * 1024 * 1024;
-    }
+    /* Reuse framebuffer that was set up by Cromwell firmware */
+    FrameBuffer.QuadPart = *((PULONG)((ULONG_PTR)DeviceExtension->VirtControlStart + NV2A_CONTROL_FRAMEBUFFER_ADDRESS_OFFSET));
+    /* Validate framebuffer address (high 4 MB of either 64 or 128 MB RAM) */
+    ASSERT(FrameBuffer.QuadPart == 0x3C00000 || FrameBuffer.QuadPart == 0x7C00000);
```

Comment:
Why do we need to assert it's at either 0x3c000000 or 0x7c000000 ? Couldn't the firmware choose to place the framebuffer completely elsewhere?

Reply:
In theory firmware can place framebuffer elsewhere, but since we don't have OS <=> firmware communication mechanism yet, we can't know the exact size of allocated range if it's non-standard... I replaced this with warning.

Reply:
Well I would say, we use whatever result returned by the firmware (but one of the assumptions that should hold is that the framebuffer should be somehow page-aligned ?)
Also in FreeLdr this area of the framebuffer should be marked as FirmwarePermanent.

Reply:
> Also in FreeLdr this area of the framebuffer should be marked as FirmwarePermanent.

It's already done. :slightly_smiling_face: 

> but one of the assumptions that should hold is that the framebuffer should be somehow page-aligned?

Something like... `ASSERT(FrameBuffer.QuadPart % PAGE_SIZE == 0);` ?

Reply:
Something like that, yes.

Reply:
Done. :slightly_smiling_face: